### PR TITLE
GH-37703: [Java] Method for setting exact number of records in ListVector

### LINKED ISSUE #37703: [Java] Add ListVector#setInitialCapacity overload that takes an exact inner capacity
### Describe the enhancement requested

ListVector (or really, its abstract base class) adds an [overload of `setInitialCapacity`](https://github.com/apache/arrow/blob/15a8ac3ce4e3ac31f9f361770ad4a38c69102aa1/java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java#L168-L206) that also preallocates the inner vector based on the "density" of the list items. However sometimes you know the number of inner elements up front, so it would be clearer (and less error-prone) to just tell the vector this fact. 

You could do this by manipulating the inner vector directly but given we have the other overload, this one also makes sense.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by jduo**
### Rationale for this change
There is currently a setInitialCapacity() function that can be used to set a number of records and density factor when setting the capacity on a ListVector. A developer may want to specify the exact total number of records instead and can use the new methods introduced here.

### What changes are included in this PR?

Add setInitialTotalCapacity() to BaseRepeatedVector, ListVector, DensityAwareVector, and LargeListVector to specify the exact total number of records in the backing vector.

This is an alternative to using the density argument in setInitialCapacity() that allows the caller to precisely specify the capacity.

### Are these changes tested?
Yes.

### Are there any user-facing changes?
No.

* Closes: #37703

### CODE REVIEW

**Comment by jduo** [comment]
@lidavidm , this looks to set the capacity on the backing vector, however I'm not sure how to verify this correctly. I'm also not sure if related fields on the outer vector are getting set correctly (getValueCapacity() on the outer vector is just returning zero after this change).

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 1]
```diff
@@ -1115,6 +1115,14 @@ public void testIsEmpty() {
     }
   }
 
+  @Test
+  public void testTotalCapacity() {
+    try (final ListVector vector = ListVector.empty("list", allocator)) {
+      vector.setInitialTotalCapacity(10, 100);
+      assertEquals(100, vector.getValueCapacity());
```
Comment (lidavidm) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java:1122]:
```suggestion
      assertEquals(10, vector.getValueCapacity());
      assertEquals(10, vector.getDataVector().getValueCapacity());
```

**Review by lidavidm — COMMENTED** [review]
The problem is a different quirk of ListVector: its child type is set _dynamically_ as you append non-null elements, and to begin with, its child is a ZeroVector which always has a zero size. Unfortunately, it seems there's no way to override this when creating a ListVector; it always delegates to a constructor of BaseRepeatedValueVector that hardcodes a ZeroVector as the child.

That explains `getDataVector().getValueCapacity()`, but not the ListVector itself. I think that's because `setInitialTotalCapacity` doesn't actually do the preallocation it promises: you have to call `allocateNewSafe()` afterwards to actually _do_ the allocation. (Also, because the allocator rounds up, you'll get more elements than expected.)

I think something like this will do it:

```
    final FieldType type = FieldType.nullable(MinorType.INT.getType());
    try (final ListVector vector = new ListVector("list", allocator, type, null)) {
      // Force the child vector to be allocated based on the type (this is a bad API: we have to track and repeat the type twice)
      vector.addOrGetVector(type);
      // Specify the allocation size but do not actually allocate
      vector.setInitialTotalCapacity(10, 100);
      // Finally actually do the allocation
      vector.allocateNewSafe();
      // Probably want to assert >= 10 and >= 100 instead of this
      assertEquals(15, vector.getValueCapacity());
      assertEquals(124, vector.getDataVector().getValueCapacity());
    }
```

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java (hunk 1 above)
Comment (jduo) [reply java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: I revised the test as suggested in the comment that demonstrates use of allocateNewSafe().

**Review by lidavidm — APPROVED** [review]
Thanks!

**Pull request merged**